relay-server/src/services/projects/cache/service.rs

```diff
@@ -0,0 +1,221 @@
+use std::sync::Arc;
+
+use relay_base_schema::project::ProjectKey;
+use relay_config::Config;
+use relay_statsd::metric;
+use relay_system::Service;
+use tokio::sync::{broadcast, mpsc};
+
+use crate::services::projects::cache::handle::ProjectCacheHandle;
+use crate::services::projects::cache::state::{CompletedFetch, Fetch, ProjectStore};
+use crate::services::projects::project::ProjectState;
+use crate::services::projects::source::ProjectSource;
+use crate::statsd::RelayTimers;
+
+/// Size of the broadcast channel for project events.
+///
+/// This is set to a value which theoretically should never be reachable,
+/// the number of events is approximately bounded by the amount of projects
+/// receiving events.
+///
+/// It is set to such a large amount because receivers of events currently
+/// do not deal with lags in the channel gracefully.
+const PROJECT_EVENTS_CHANNEL_SIZE: usize = 512_000;
+
+/// A cache for projects, which allows concurrent access to the cached projects.
+#[derive(Debug)]
+pub enum ProjectCache {
+    /// Schedules another fetch or update for the specified project.
+    ///
+    /// A project which is not fetched will eventually expire and be evicted
+    /// from the cache. Fetches for an already cached project ensure the project
+    /// is always up to date and not evicted.
+    Fetch(ProjectKey),
+}
+
+impl ProjectCache {
+    fn variant(&self) -> &'static str {
+        match self {
+            Self::Fetch(_) => "fetch",
+        }
+    }
+}
+
+impl relay_system::Interface for ProjectCache {}
+
+impl relay_system::FromMessage<Self> for ProjectCache {
+    type Response = relay_system::NoResponse;
+
+    fn from_message(message: Self, _: ()) -> Self {
+        message
+    }
+}
+
+/// Project life-cycle events produced by the project cache.
+#[derive(Debug, Copy, Clone)]
+pub enum ProjectEvent {
```

Comment:
nit: I think `Event` is a good name for this in general, but it has a special meaning in sentry context. Should we call it `ProjectSignal` or something like that?